packages/client/src/index.ts

```diff
@@ -2,15 +2,80 @@ export * from './error';
+ *
+ * Any script in page scope may create an object like this, so clients should
+ * confirm a provider is actually present by confirming provider origin and
+ * fetching the provider manifest. Provider details such as name, version,
+ * website, brief descriptive text, and icons should be available in the
+ * manifest.
+ * 
+ * Presently clients may expect the manifest is a chrome extension manifest v3.
+ * @see https://developer.chrome.com/docs/extensions/reference/manifest
+ *
+ * Clients may `request()` approval to connect. This method may reject if the
+ * provider chooses to deny approval.  Approval granted by a successful
+ * request will persist accross sessions.
+ *
+ * Clients must `connect()` to acquire a `MessagePort`. The resulting
+ * `MessagePort` represents an active, type-safe communication channel to the
+ * provider. It is convenient to provide the `connect` method as the `getPort`
+ * option for `createChannelTransport` from `@penumbra-zone/transport-dom`, or
+ * use the helpers available in `@penumbra-zone/client/create`.
+ * 
+   ```js
+   import { jsonOptions } from '@penumbra-zone/protobuf';
+   import { createChannelTransport } from '@penumbra-zone/transport-dom';
+
+   // naively get first available provider
+   const provider = Object.values(window[Symbol.for('penumbra')])[0];
+   void provider.request();
+
+   // establish a transport
+   const transport = createChannelTransport({ jsonOptions, getPort: provider.connect });
+
+   // export function to create client
+   export const createPenumbraClient = serviceType => createPromiseClient(serviceType, transport);
+   ```
+*
+* 
+*/
 export interface PenumbraInjection {
+  /** Should be called by a page creating a channel transport. It returns a
+   * promise that may resolve with an active `MessagePort` providing a channel
+   * to this provider. */
   readonly connect: () => Promise<MessagePort>;
+
+  /** Should be called by a page that does not yet have approval to connect.
+   * Returns a promise that may reject with an enumerated failure reason. */
   readonly request: () => Promise<void>;
+
+  /** Call to indicate the provider should revoke approval of this origin. */
+  readonly disconnect: () => Promise<void>;
+
+  /** Should synchronously return the present connection state.
+   *
+   * - `true` indicates active connection.
+   * - `false` indicates connection is closed or rejected.
+   * - `undefined` indicates connection may be attempted.
+   */
   readonly isConnected: () => boolean | undefined;
+
+  /** Should contain a URI located at the provider's origin, which returns a
+   * chrome extension manifest v3 describing this provider. */
   readonly manifest: string;
 }
 
 declare global {
   interface Window {
+    /** Records upon this global should identify themselves by a field name
+     * matching the origin of the provider. */
     readonly [PenumbraSymbol]?: undefined | Readonly<Record<string, PenumbraInjection>>;
```

Comment:
question: can we export some function like `getInjectedWallet(walletOrigin)` instead of using `PenumbraSymbol`?

I imagine external users implement the same API for wallets, so a set of function could help them:
- `getInjectedWallet(walletOrigin)` – for dApp devs
- `getInjectedWallets()` – list all browser wallets
- `injectWallet({ connect, disconnect, onConnectionChange ... })` – for wallet developers
- `onWalletInjected(callback)` – a function that fires a callback each time a wallet is added (no need to wait for page reload). For this `window[PenumbraSymbol]` should probably be converted to a proxy object to control additions and deletions, idk.

`getInjectedWallet` could become the beginning of this improved DX

Reply:
see utils in create.ts

other aspects of this should probably happen after externally_connectable

injection should generally happen before page scripts load unless not approved.

adding callback functionality seems less useful, as the wallet client and all requests will be entirely promise-based anyway

react wrapper in another PR implements context and hooks, for that style

Reply:
> injection should generally happen before page scripts load unless not approved

agree except for the case when user installs the extension and the page is loaded. Real life scenario:
1. user enters the frontend
2. frontend says "to get started, install the extension"
3. they install the extension
4. page didn't react – for now

i know it's a small thing but, in my opinion, developers will find the `onWalletInjected(callback)`

Reply:
the page must reload before a new content script can appear.

Reply:
react wrapper in https://github.com/penumbra-zone/web/pull/1293

and another style, connect/query use, is visible in https://github.com/penumbra-zone/nextjs-penumbra-client-example/commit/228b6eba37a2b2419e8e5e3d0b8510c20a3a9ede

and the new readme for this package https://github.com/penumbra-zone/web/pull/1307/files#diff-7b3f04edbf937e50b69972d84ca0e92c9078c962fb4965bb7647b7c6ba74c296